st2client/st2client/commands/auth.py

```diff
@@ -70,6 +73,124 @@ def run_and_print(self, args, **kwargs):
+
+        kwargs['has_token_opt'] = False
+
+        super(LoginCommand, self).__init__(
+            resource, kwargs.pop('name', 'create'),
+            'Authenticate user, acquire access token, and update CLI config directory',
+            *args, **kwargs)
+
+        self.parser.add_argument('username',
+                                 help='Name of the user to authenticate.')
+
+        self.parser.add_argument('-p', '--password', dest='password',
+                                 help='Password for the user. If password is not provided, '
+                                      'it will be prompted.')
+        self.parser.add_argument('-l', '--ttl', type=int, dest='ttl', default=None,
+                                 help='The life span of the token in seconds. '
+                                      'Max TTL configured by the admin supersedes this.')
+        self.parser.add_argument('-w', '--write-password', action='store_true', default=False,
+                                 dest='write_password',
+                                 help='Write the password in plain text to the config file '
+                                      '(default is to omit it')
+
+    def run(self, args, **kwargs):
+
+        if not args.password:
+            args.password = getpass.getpass()
+        instance = self.resource(ttl=args.ttl) if args.ttl else self.resource()
+
+        cli = BaseCLIApp()
+
+        # Determine path to config file
+        try:
+            config_file = cli._get_config_file_path(args)
+        except ValueError:
+            # config file not found in args or in env, defaulting
+            config_file = config_parser.ST2_CONFIG_PATH
+
+        # Retrieve token
+        manager = self.manager.create(instance, auth=(args.username, args.password), **kwargs)
+        cli._cache_auth_token(token_obj=manager)
+
+        # Update existing configuration with new credentials
+        config = ConfigParser()
+        config.read(config_file)
+
+        # Modify config (and optionally populate with password)
+        if 'credentials' not in config:
+            config.add_section('credentials')
+            config['credentials'] = {}
+        config['credentials']['username'] = args.username
+        if args.write_password:
+            config['credentials']['password'] = args.password
+        else:
+            # Remove any existing password from config
+            config['credentials'].pop('password', None)
+
+        with open(config_file, "w") as cfg_file_out:
+            config.write(cfg_file_out)
+
+        return manager
+
+    def run_and_print(self, args, **kwargs):
+        try:
+            self.run(args, **kwargs)
+            print("Logged in as %s" % (args.username))
+        except Exception as e:
+            print("Failed to log in as %s: %s" % (args.username, str(e)))
+            if self.app.client.debug:
+                raise
+
+
+class WhoamiCommand(resource.ResourceCommand):
```

Comment:
@Kami Okay, I tested both the ST2_CONFIG_PATH` and `--config-file` options, and other than running into [this bug](
https://github.com/StackStorm/st2/issues/3186) (unrelated to this patch) everything seems to work using a totally different config path.

It should be noted that this tool doesn't actually create the configuration file - as with other commands, if the file doesn't exist at the referenced path, it will raise an exception. I did this to be as predictable compared to other st2 commands as possible.